File: tree.js
const DFS = 0;    // Depth First Search / Stack
const BFS = 1;    // Breadth First Search / Queue
const RDFS = 2;   // Reversed DFS

class Queue {
    constructor() { this._arr = []; }
    enqueue(item) { this._arr.push(item); }
    dequeue() { return this._arr.shift(); }
    empty() { return (this._arr.length == 0)? true:false; }
}

class Stack {
    constructor() { this._arr = []; }
    push(item) { this._arr.push(item); }
    pop() { return this._arr.pop(); }
    empty() { return (this._arr.length == 0)? true:false; }
}

class Utils {
    static pad(n, width, c) {
        n = n + '';
        return n.length >= width ? n : n + new Array(width - n.length + 1).join(c);
    }
}

class Tree {
    constructor(data=null) {
        this.nodes = null;
        this.root = null;
        this.method = null
        this.queue = null;
        this.stack = null;
        this.visited = null;
        if (data) {
            this.loadData(data);
        }
        this.setMethod('DFS');
    }

    loadData(data) {
        this.nodes = {};
        for (var i in data) {
            var key = data[i]['id'];
            this.nodes[key] = data[i];
            if (data[i]['parent'] == 'NULL')
                this.root = key;
        }
    }

    setMethod(method) {
        if (method == 'DFS')
            this.method = DFS;
        if (method == 'BFS')
            this.method = BFS;
        if (method == 'RDFS')
            this.method = RDFS;
        delete this.queue;
        delete this.stack;
        delete this.visited;
        this.queue = new Queue();
        this.stack = new Stack();
        this.visited = {};
    }

    start() {
        if (this.method == DFS)
            this.stack.push(this.root);
        if (this.method == BFS)
            this.queue.enqueue(this.root);
        if (this.method == RDFS)
            this.stack.push(this.root);
        return this.nodes[this.root];
    }

    next(callback=null) {
        if (this.method == DFS) {
            if (!this.stack.empty()) {
                var id = this.stack.pop();
                var node = this.nodes[id];
                if (callback) callback(node);
                if (node['child']) {
                    var childs = node['child'].split(',');
                    for (var i = childs.length - 1; i >= 0; i--)
                        this.stack.push(childs[i]);
                }
                return node;
            } else return null;
        }

        if (this.method == BFS) {
            if (!this.queue.empty()) {
                var id = this.queue.dequeue();
                var node = this.nodes[id];
                if (callback) callback(node);
                if (node['child']) {
                    var childs = node['child'].split(',');
                    for (var i = 0; i < childs.length; i++)
                        this.queue.enqueue(childs[i]);
                }
                return node;
            } else return null;
        }

        if (this.method == RDFS) {
            if (!this.stack.empty()) {
                var id = this.stack.pop();
                var node = this.nodes[id];
                var childs = new Array();
                if (node['child']) {
                    childs = node['child'].split(',');
                    for (var i = childs.length - 1; i >= 0; i--)
                        if (this.visited[childs[i]])
                            childs = childs.filter(function(value, index, arr){ return value != childs[i]; });
                }
                if (childs.length > 0) {
                    this.stack.push(id);
                    for (var i = childs.length - 1; i >= 0; i--)
                        this.stack.push(childs[i]);
                    return 'skipped';
                } else {
                    if (callback) callback(node);
                    this.visited[id] = 1;
                    return node;
                }
            } else return null;
        }
    }
}

File: tree.py
# -*- coding: utf-8 -*-
import constant
from collections import deque


class Tree:
    def __init__(self, data=None):
        self.nodes = None
        self.root = None
        self.method = constant.DFS
        self.queue = deque()
        self.stack = deque()
        self.visited = {}
        if data:
            self.loadData(data)
        self.setMethod('DFS');
        return

    def loadData(self, data):
        self.nodes = {}
        for node in data:
            key = node['id']
            self.nodes[key] = node
            if node['parent'] == 'NULL':
                self.root = key

    def setMethod(self, method):
        if method == 'DFS':
            self.method = constant.DFS
        if method == 'BFS':
            self.method = constant.BFS
        if method == 'RDFS':
            self.method = constant.RDFS
        self.queue.clear()
        self.stack.clear()
        self.visited = {}

    def start(self):
        if self.method == constant.DFS:
            self.stack.append(self.root)
        if self.method == constant.BFS:
            self.queue.append(self.root)
        if self.method == constant.RDFS:
            self.stack.append(self.root)
        return self.nodes[self.root]

    def next(self, callback=None):
        if self.method == constant.DFS:
            if self.stack:
                id = self.stack.pop()
                node = self.nodes[id]
                if callback:
                    callback(node)
                if node['child']:
                    childs = node['child'].split(",")
                    for key in reversed(childs):
                        self.stack.append(key)
                return node
            else:
                return None

        if self.method == constant.BFS:
            if self.queue:
                id = self.queue.popleft()
                node = self.nodes[id]
                if callback:
                    callback(node)
                if node['child']:
                    childs = node['child'].split(",")
                    for key in childs:
                        self.queue.append(key)
                return node
            else:
                return None

        if self.method == constant.RDFS:
            if self.stack:
                id = self.stack.pop()
                node = self.nodes[id]
                childs = None
                if node['child']:
                    childs = node['child'].split(",")
                    for key in reversed(childs):
                        if key in self.visited:
                            childs.remove(key)
                if childs:
                    self.stack.append(id)
                    for key in reversed(childs):
                        self.stack.append(key)
                    return 'skipped'
                else:
                    if callback:
                        callback(node)
                    self.visited[id] = 1
                    return node
            else:
                return None

